File: android-app/app/src/main/java/com/lasohealth/android/core/analysis/analyzers/InsightCoordinator.kt
package com.lasohealth.android.core.analysis.analyzers

import com.lasohealth.android.core.analysis.AnalysisContext
import com.lasohealth.android.core.analysis.AnalysisInsight
import com.lasohealth.android.core.analysis.InsightAnalyzer
import com.lasohealth.android.core.analysis.InsightDirective
import com.lasohealth.android.core.analysis.InsightPriority
import com.lasohealth.android.core.model.HealthMetric
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.coroutineScope

/**
 * Coordinates all insight analyzers: runs them, resolves contradictions,
 * deduplicates, and caps the result set.
 *
 * Resolution order: safety > severity > evidence depth.
 *
 * Pipeline:
 *   1. Run all analyzers (concurrently via coroutine fan-out)
 *   2. Infer directives for insights still tagged INFORMATIONAL (where possible)
 *   3. Resolve contradictions (e.g. REST vs INCREASE_ACTIVITY) — safety wins
 *   4. Deduplicate by metric (max 2 per metric)
 *   5. Sort by priority and cap at [MAX_INSIGHTS]
 *
 * Ported from iOS InsightCoordinator.swift.
 */
class InsightCoordinator {

    /**
     * Run all [analyzers] against the given [context], resolve contradictions,
     * deduplicate, and return at most [MAX_INSIGHTS] prioritised insights.
     */
    suspend fun coordinate(
        context: AnalysisContext,
        analyzers: List<InsightAnalyzer>,
    ): List<AnalysisInsight> = coroutineScope {
        // Step 1: Run all analyzers concurrently
        val rawInsights = analyzers
            .map { analyzer ->
                async {
                    try {
                        analyzer.generateInsights(context)
                    } catch (_: Exception) {
                        emptyList()
                    }
                }
            }
            .awaitAll()
            .flatten()

        if (rawInsights.isEmpty()) return@coroutineScope emptyList()

        // Step 2: Infer directives for any insight still tagged INFORMATIONAL
        var tagged = rawInsights.map { insight ->
            if (insight.directive == InsightDirective.INFORMATIONAL) {
                insight.copy(directive = inferDirective(insight))
            } else {
                insight
            }
        }

        // Step 3: Resolve contradictions
        tagged = resolveConflicts(tagged)

        // Step 4: Deduplicate (max 2 per metric)
        tagged = deduplicateByMetric(tagged)

        // Step 5: Sort by priority and cap
        tagged
            .sortedWith(
                compareBy<AnalysisInsight> { it.priority.ordinal }
                    .thenByDescending { it.confidence },
            )
            .take(MAX_INSIGHTS)
    }

    // ----- Directive Inference -----

    /**
     * Infer a directive from the insight's title and detail text when the analyzer
     * did not assign one explicitly.
     */
    private fun inferDirective(insight: AnalysisInsight): InsightDirective {
        val title = insight.title.lowercase()
        val detail = insight.detail.lowercase()

        // Text-based rules ordered by priority: medical > rest > sleep > activity > intensity
        for (rule in TEXT_RULES) {
            if (rule.matches(title, detail)) {
                return rule.directive
            }
        }

        return InsightDirective.INFORMATIONAL
    }

    // ----- Conflict Resolution -----

    /**
     * Resolve contradicting insights. When two insights have conflicting directives
     * and at least one is HIGH priority or above, the safer/more important directive wins.
     */
    private fun resolveConflicts(insights: List<AnalysisInsight>): List<AnalysisInsight> {
        val suppressed = mutableSetOf<String>()

        for (insightA in insights) {
            if (insightA.id in suppressed) continue
            val conflictsA = CONFLICTS[insightA.directive] ?: continue
            if (conflictsA.isEmpty()) continue

            for (insightB in insights) {
                if (insightA.id == insightB.id) continue
                if (insightB.id in suppressed) continue
                if (insightB.directive !in conflictsA) continue

                // Only resolve if at least one is HIGH+ priority
                if (insightA.priority.ordinal > InsightPriority.HIGH.ordinal &&
                    insightB.priority.ordinal > InsightPriority.HIGH.ordinal
                ) continue

                // Determine winner
                val priorityA = RESOLUTION_PRIORITY[insightA.directive] ?: 0
                val priorityB = RESOLUTION_PRIORITY[insightB.directive] ?: 0

                val aWins = when {
                    priorityA != priorityB -> priorityA > priorityB
                    insightA.priority != insightB.priority ->
                        insightA.priority.ordinal < insightB.priority.ordinal
                    else -> insightA.confidence >= insightB.confidence
                }

                suppressed += if (aWins) insightB.id else insightA.id

                // If A lost, stop checking A against others
                if (!aWins) break
            }
        }

        return insights.filter { it.id !in suppressed }
    }

    // ----- Deduplication -----

    /**
     * Remove duplicate insights about the same metric across all analyzers.
     * Keeps max 2 per metric, selecting by priority then confidence.
     */
    private fun deduplicateByMetric(insights: List<AnalysisInsight>): List<AnalysisInsight> {
        val grouped = insights.groupBy { it.metric }
        val result = mutableListOf<AnalysisInsight>()

        for ((_, metricInsights) in grouped) {
            if (metricInsights.size <= 2) {
                result += metricInsights
                continue
            }

            // Sort by priority (critical first), then confidence descending
            val sorted = metricInsights.sortedWith(
                compareBy<AnalysisInsight> { it.priority.ordinal }
                    .thenByDescending { it.confidence },
            )

            // Keep the top 2
            result += sorted.take(2)
        }

        return result
    }

    // ----- Conflict Tables -----

    private data class TextRule(
        val directive: InsightDirective,
        val titleKeywords: List<String>,
        val detailKeywords: List<String>,
    ) {
        fun matches(title: String, detail: String): Boolean {
            val titleMatch = titleKeywords.isEmpty() || titleKeywords.any { title.contains(it) }
            val detailMatch = detailKeywords.isEmpty() || detailKeywords.any { detail.contains(it) }
            return when {
                titleKeywords.isEmpty() -> detailMatch
                detailKeywords.isEmpty() -> titleMatch
                else -> titleMatch || detailMatch
            }
        }
    }

    companion object {
        const val MAX_INSIGHTS = 15

        /**
         * Directives that directly contradict each other.
         * Two insights with mutually conflicting directives cannot coexist
         * when at least one carries HIGH+ priority.
         */
        private val CONFLICTS: Map<InsightDirective, Set<InsightDirective>> = mapOf(
            InsightDirective.REST to setOf(
                InsightDirective.INCREASE_ACTIVITY,
                InsightDirective.PUSH_HARDER,
            ),
            InsightDirective.INCREASE_ACTIVITY to setOf(
                InsightDirective.REST,
                InsightDirective.REDUCE_INTENSITY,
            ),
            InsightDirective.REDUCE_INTENSITY to setOf(
                InsightDirective.PUSH_HARDER,
                InsightDirective.INCREASE_ACTIVITY,
            ),
            InsightDirective.PUSH_HARDER to setOf(
                InsightDirective.REST,
                InsightDirective.REDUCE_INTENSITY,
            ),
            InsightDirective.SLEEP_MORE to emptySet(),
            InsightDirective.SLEEP_BETTER to emptySet(),
            InsightDirective.SEEK_MEDICAL to emptySet(),     // never suppressed
            InsightDirective.MAINTAIN to emptySet(),
            InsightDirective.INFORMATIONAL to emptySet(),
        )

        /**
         * Resolution priority for conflicting directives.
         * Higher value wins. Safety-oriented directives outrank performance pushes.
         */
        private val RESOLUTION_PRIORITY: Map<InsightDirective, Int> = mapOf(
            InsightDirective.SEEK_MEDICAL to 100,
            InsightDirective.REST to 80,
            InsightDirective.REDUCE_INTENSITY to 70,
            InsightDirective.SLEEP_MORE to 60,
            InsightDirective.SLEEP_BETTER to 50,
            InsightDirective.INCREASE_ACTIVITY to 40,
            InsightDirective.PUSH_HARDER to 30,
            InsightDirective.MAINTAIN to 20,
            InsightDirective.INFORMATIONAL to 10,
        )

        /**
         * Text-based rules for inferring directives from insight content.
         * Ordered by priority: medical > rest > sleep > activity > intensity.
         */
        private val TEXT_RULES = listOf(
            TextRule(
                InsightDirective.SEEK_MEDICAL,
                emptyList(),
                listOf("consult", "healthcare provider", "doctor"),
            ),
            TextRule(
                InsightDirective.REST,
                listOf("overtraining", "rest day", "recovery day", "rest deficit"),
                listOf("rest day", "genuine rest"),
            ),
            TextRule(
                InsightDirective.SLEEP_MORE,
                listOf("sleep debt"),
                listOf("bedtime alarm", "more sleep"),
            ),
            TextRule(
                InsightDirective.SLEEP_BETTER,
                listOf("sleep consistency", "bedtime"),
                listOf("consistent bedtime"),
            ),
            TextRule(
                InsightDirective.INCREASE_ACTIVITY,
                listOf("workout consistency"),
                listOf("20-min walk", "20 min walk", "more active"),
            ),
            TextRule(
                InsightDirective.PUSH_HARDER,
                emptyList(),
                listOf("push harder", "increase intensity", "progressive overload"),
            ),
            TextRule(
                InsightDirective.REDUCE_INTENSITY,
                emptyList(),
                listOf("lower intensity", "lighter sessions", "reduce intensity"),
            ),
        )
    }
}

File: Analysis/InsightCoordinator.swift
import Foundation

/// Resolves contradictions, deduplicates, and caps insights from 20+ independent analyzers.
/// Resolution order: safety > severity > evidence depth.
struct InsightCoordinator {

    private static let maxInsights = 15

    static func coordinate(_ insights: [Insight]) -> [Insight] {
        guard !insights.isEmpty else { return [] }

        // Step 1: Infer directives for any insight still tagged .informational
        var tagged = insights.map { insight -> Insight in
            var copy = insight
            if copy.directive == .informational {
                copy.directive = inferDirective(for: copy)
            }
            return copy
        }

        // Step 2: Resolve contradictions
        tagged = resolveConflicts(tagged)

        // Step 3: Deduplicate (max 2 per metric)
        tagged = deduplicateByMetric(tagged)

        // Step 4: Sort by priority and cap
        tagged.sort { $0.priorityScore > $1.priorityScore }
        return Array(tagged.prefix(maxInsights))
    }

    // MARK: - Directive Inference

    static func inferDirective(for insight: Insight) -> InsightDirective {
        // Category-level fast paths
        switch insight.category {
        case .illnessWarning: return .rest
        case .clinicalTrajectory: return insight.severity >= .warning ? .seekMedical : .informational
        case .personalRecord: return .maintain
        case .cyclePhase: return .informational
        case .multiMetricCluster where insight.severity >= .warning: return .rest
        case .recovery where insight.severity >= .warning: return .rest
        default: break
        }

        // Cognitive/energy sub-types
        if insight.category == .cognitiveEnergy {
            if let d = cognitiveDirective(for: insight) { return d }
        }

        let title = insight.title.lowercased()
        let rec = insight.recommendation.lowercased()

        // Text-based rules (ordered by priority: medical > rest > sleep > activity > intensity)
        for rule in Self.textRules {
            if rule.matches(title: title, rec: rec, trend: insight.trend) {
                return rule.directive
            }
        }

        if insight.trend == .improving && insight.severity == .info { return .maintain }
        return .informational
    }

    // MARK: - Text Rule Table

    private struct TextRule {
        let directive: InsightDirective
        let titleKeywords: [String]
        let recKeywords: [String]
        let requiredTrend: TrendDirection?

        func matches(title: String, rec: String, trend: TrendDirection?) -> Bool {
            let titleMatch = titleKeywords.isEmpty || titleKeywords.contains(where: { title.contains($0) })
            let recMatch = recKeywords.isEmpty || recKeywords.contains(where: { rec.contains($0) })
            let trendMatch = requiredTrend == nil || trend == requiredTrend
            return (titleKeywords.isEmpty ? recMatch : recKeywords.isEmpty ? titleMatch : titleMatch || recMatch) && trendMatch
        }
    }

    private static let textRules: [TextRule] = [
        TextRule(directive: .seekMedical, titleKeywords: [], recKeywords: ["consult", "healthcare provider", "doctor"], requiredTrend: nil),
        TextRule(directive: .rest, titleKeywords: ["overtraining", "rest day", "recovery day", "rest deficit"], recKeywords: ["rest day", "genuine rest"], requiredTrend: nil),
        TextRule(directive: .sleepMore, titleKeywords: ["sleep debt"], recKeywords: ["bedtime alarm", "more sleep"], requiredTrend: nil),
        TextRule(directive: .sleepBetter, titleKeywords: ["sleep consistency", "bedtime"], recKeywords: ["consistent bedtime"], requiredTrend: nil),
        TextRule(directive: .increaseActivity, titleKeywords: ["workout consistency"], recKeywords: [], requiredTrend: .declining),
        TextRule(directive: .increaseActivity, titleKeywords: [], recKeywords: ["20-min walk", "20 min walk", "more active"], requiredTrend: nil),
        TextRule(directive: .pushHarder, titleKeywords: [], recKeywords: ["push harder", "increase intensity", "progressive overload"], requiredTrend: nil),
        TextRule(directive: .reduceIntensity, titleKeywords: [], recKeywords: ["lower intensity", "lighter sessions", "reduce intensity"], requiredTrend: nil),
    ]

    private static func cognitiveDirective(for insight: Insight) -> InsightDirective? {
        let title = insight.title.lowercased()
        if title.contains("mental fatigue") || title.contains("recovery day") { return .rest }
        if (title.contains("sleep debt") || title.contains("cognitive readiness")) && insight.trend == .declining { return .sleepMore }
        if title.contains("physical energy") { return .increaseActivity }
        return nil
    }

    // MARK: - Conflict Resolution

    private static func resolveConflicts(_ insights: [Insight]) -> [Insight] {
        var suppressed = Set<UUID>()

        // Build directive lookup
        let directiveMap: [UUID: InsightDirective] = Dictionary(
            uniqueKeysWithValues: insights.map { ($0.id, $0.directive) }
        )

        for insightA in insights {
            guard !suppressed.contains(insightA.id) else { continue }
            let directiveA = directiveMap[insightA.id] ?? .informational
            guard !directiveA.conflicts.isEmpty else { continue }

            for insightB in insights {
                guard insightA.id != insightB.id else { continue }
                guard !suppressed.contains(insightB.id) else { continue }
                let directiveB = directiveMap[insightB.id] ?? .informational
                guard directiveA.conflicts.contains(directiveB) else { continue }

                // Determine winner
                let aWins: Bool
                if directiveA.resolutionPriority != directiveB.resolutionPriority {
                    aWins = directiveA.resolutionPriority > directiveB.resolutionPriority
                } else if insightA.severity != insightB.severity {
                    aWins = insightA.severity > insightB.severity
                } else {
                    aWins = insightA.priorityScore >= insightB.priorityScore
                }

                suppressed.insert(aWins ? insightB.id : insightA.id)

                // If A lost, stop checking A against others
                if !aWins { break }
            }
        }

        return insights.filter { !suppressed.contains($0.id) }
    }

    // MARK: - Deduplication

    /// Remove duplicate insights about the same metric across all analyzers.
    /// Keeps max 2 per metric: best causal-chain + best other (by priority score).
    private static func deduplicateByMetric(_ insights: [Insight]) -> [Insight] {
        var grouped: [HealthMetric: [Insight]] = [:]
        for insight in insights {
            grouped[insight.metric, default: []].append(insight)
        }

        var result: [Insight] = []
        for (_, metricInsights) in grouped {
            if metricInsights.count <= 1 {
                result.append(contentsOf: metricInsights)
                continue
            }

            let causalChains = metricInsights.filter { $0.category == .causalChain }
            let others = metricInsights.filter { $0.category != .causalChain }

            // Keep the best causal chain if any
            if let bestCausal = causalChains.max(by: { $0.priorityScore < $1.priorityScore }) {
                result.append(bestCausal)
            }

            // Keep the single best non-causal insight
            if let bestOther = others.max(by: { $0.priorityScore < $1.priorityScore }) {
                result.append(bestOther)
            }
        }

        return result
    }
}
